challenger/src/serializing_challenger.rs

```diff
@@ -0,0 +1,228 @@
+
+use p3_field::{ExtensionField, PrimeField32, PrimeField64};
+use p3_maybe_rayon::prelude::*;
+use p3_symmetric::{CryptographicHasher, Hash};
+use p3_util::log2_ceil_u64;
+use tracing::instrument;
+
+use crate::{
+    CanObserve, CanSample, CanSampleBits, FieldChallenger, GrindingChallenger, HashChallenger,
+};
+
+/// Given a challenger that can observe and sample bytes, produces a challenger that is able to
+/// sample and observe field elements of a `PrimeField32`.
+///
+/// **Observing**:
+/// -  Takes a field element will serialize it into a byte array and observe each byte.
+///
+/// **Sampling**:
+/// -  Samples a field element in a prime field of size `p` by sampling unofrmly an element in the
+///    range (0..1 << log_2(p)). This avoids modulo bias.
+#[derive(Clone, Debug)]
+pub struct SerializingChallenger32<F, Inner> {
+    inner: Inner,
+    _marker: PhantomData<F>,
+}
+
+/// Given a challenger that can observe and sample bytes, produces a challenger that is able to
+/// sample and observe field elements of a `PrimeField64` field.
+///
+/// **Observing**:
+/// -  Takes a field element will serialize it into a byte array and observe each byte.
+///
+/// **Sampling**:
+/// -  Samples a field element in a prime field of size `p` by sampling unofrmly an element in the
+///    range (0..1 << log_2(p)). This avoids modulo bias.
+#[derive(Clone, Debug)]
+pub struct SerializingChallenger64<F, Inner> {
+    inner: Inner,
+    _marker: PhantomData<F>,
+}
+
+impl<F: PrimeField32, Inner: CanObserve<u8>> SerializingChallenger32<F, Inner> {
+    pub fn new(inner: Inner) -> Self {
+        Self {
+            inner,
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<F, H> SerializingChallenger32<F, HashChallenger<u8, H, 32>>
+where
+    F: PrimeField32,
+    H: CryptographicHasher<u8, [u8; 32]>,
+{
+    pub fn from_hasher(initial_state: Vec<u8>, hasher: H) -> Self {
+        Self::new(HashChallenger::new(initial_state, hasher))
+    }
+}
+
+impl<F: PrimeField32, Inner: CanObserve<u8>> CanObserve<F> for SerializingChallenger32<F, Inner> {
+    fn observe(&mut self, value: F) {
+        self.inner
+            .observe_slice(&value.as_canonical_u32().to_le_bytes());
+    }
+}
+
+impl<F: PrimeField32, const N: usize, Inner: CanObserve<u8>> CanObserve<Hash<F, u8, N>>
+    for SerializingChallenger32<F, Inner>
+{
+    fn observe(&mut self, values: Hash<F, u8, N>) {
+        for value in values {
+            self.inner.observe(value);
+        }
+    }
+}
+
+impl<F, EF, Inner> CanSample<EF> for SerializingChallenger32<F, Inner>
+where
+    F: PrimeField32,
+    EF: ExtensionField<F>,
+    Inner: CanSample<u8>,
+{
+    fn sample(&mut self) -> EF {
+        let log_size = log2_ceil_u64(F::ORDER_U64).saturating_sub(1);
+        let bound = (1 << log_size) - 1;
+        let sample_base = |inner: &mut Inner| {
+            let value = u32::from_le_bytes(inner.sample_array::<4>());
+            F::from_canonical_u32(value & bound)
```

Comment:
Couldn't `value` exceed the canonical range (even though it can't reach the next power of two)? I think it's best to do rejection sampling, retrying until `value < F::ORDER_U64`

Reply:
I am sampling over 0 <= x < 2^{log(p)} so that 2^{log(p)} < p so I believe it's ok in terms of range. Here `log(p)` stands for `log_ceil(p) - 1`. 

I think you are right that we should do rejection sampling over the whole range in order to include all the elements so I changed it